File: ts/src/parse.ts
import { Prompt, PromptTemplate, Role, Roles } from "./types";
import { parse as parseYaml } from "yaml";

const frontmatterRegex = /^---\n([\s\S]*?)\n---\n/;

/**
 * parse parses the given prompt file into a Prompt object representing (optional) metadata and
 * message templates.
 *
 * @param content Input file
 * @returns Parsed prompt
 */
export function parse(content: string): Prompt {
  // Check if there is any frontmatter. Frontmatter is separated from the markdown content by `---`
  // before and after the content
  const frontMatterMatch = frontmatterRegex.exec(content);
  let metadata: { [key: string]: unknown } | undefined;

  if (frontMatterMatch) {
    const frontmatterContent = frontMatterMatch[1];
    metadata = parseYaml(frontmatterContent);

    // Remove frontmatter from prompt content
    const frontMatterEndIndex =
      frontMatterMatch.index + frontMatterMatch[0].length;
    content = content.substring(frontMatterEndIndex).trim();
  }

  const messages = parseMessages(content);

  return {
    metadata,
    messages,
  };
}

const roleRegex = new RegExp(
  `\\s*#?\\s*(${Roles.join("|")})\\s*:\\s*\\n`,
  "im"
);

function parseMessages(content: string): PromptTemplate[] {
  // Parse the template(s)
  const messages: PromptTemplate[] = [];

  // Trim and remove empty chunks
  const chunks = content.split(roleRegex).filter((x) => x.trim());

  // If it's just one chunk, default to a system message
  if (chunks.length === 1) {
    return [
      {
        role: "system",
        message: content.trim(),
      },
    ];
  }

  // Chunks should be even, as they are split by the regex
  if (chunks.length % 2 !== 0) {
    throw new Error("Invalid format");
  }

  for (let i = 0; i < chunks.length; i += 2) {
    const chunk = chunks[i];

    const role = chunk.trim() as Role;
    const message = chunks[i + 1].trim();

    messages.push({
      role,
      message,
    });
  }

  return messages;
}

File: go/parse.go
package promptmd

import (
	"regexp"
	"strings"

	"github.com/stretchr/testify/assert/yaml"
)

func Parse(content string) (*Prompt, error) {
	p := &Prompt{}

	// Parse metadata
	// Check if there is frontmatter
	if strings.HasPrefix(content, "---") {
		end := strings.Index(content[3:], "---")
		if end > 0 {
			// Parse frontmatter
			frontmatter := content[3 : end+3]
			var metadata map[string]any
			err := yaml.Unmarshal([]byte(frontmatter), &metadata)
			if err != nil {
				return nil, err
			}

			p.Metadata = metadata

			// Remove frontmatter from content
			content = content[end+6:]
		}
	}

	// Parse content
	messages, err := parseMessages(content)
	if err != nil {
		return nil, err
	}
	p.Messages = messages

	return p, nil
}

var roleRegex = regexp.MustCompile(
	`(?m)^\s*#?\s*(` + strings.Join(stringSlice(Roles), "|") + `)\s*:\s*$`,
)

func stringSlice(roles []Role) []string {
	ss := make([]string, len(roles))
	for i, r := range roles {
		ss[i] = string(r)
	}
	return ss
}

func parseMessages(content string) ([]Message, error) {
	// Parse the template(s)
	messages := []Message{}

	matches := roleRegex.FindAllStringSubmatchIndex(content, -1)

	// If there are no roles, assume the entire content is a system prompt
	if len(matches) == 0 {
		messages = append(messages, Message{
			Role:    RoleSystem,
			Message: strings.TrimSpace(content),
		})
		return messages, nil
	}

	// Iterate over the matches and extract the content
	for i, match := range matches {
		// Extract the role
		role := Role(strings.TrimSpace(content[match[2]:match[3]]))

		var contentStart int
		if i == len(matches)-1 {
			contentStart = match[1]
		} else {
			contentStart = match[1]
		}

		var contentEnd int
		if i == len(matches)-1 {
			contentEnd = len(content)
		} else {
			contentEnd = matches[i+1][0]
		}

		messageContent := content[contentStart:contentEnd]

		// Extract the content
		messages = append(messages, Message{
			Role:    role,
			Message: strings.TrimSpace(messageContent),
		})
	}

	return messages, nil
}
